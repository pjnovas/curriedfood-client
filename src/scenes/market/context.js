import React from 'react';
import { get } from 'lodash';

import { useRequest, useRequestSWR } from 'hooks/service';

const ShoppingContext = React.createContext();

const getUri = (id) => `/shopping-lists/${id || ''}`;

const useUpdate = ({ data, revalidate }) => {
  const request = useRequest();

  return async (groceries) => {
    await request({
      url: getUri(data.id),
      method: 'put',
      data: {
        groceries
      }
    });

    revalidate();
  };
};

const ShoppingProvider = ({ children }) => {
  const value = useRequestSWR({ url: getUri() }, { withPlace: true });
  const data = get(value, 'data[0]') || null;
  const update = useUpdate({ ...value, data });

  const parsed = {
    ...value,
    update,
    data,
    mutate: (groceries) => {
      value.mutate({
        ...data,
        groceries
      });
    }
  };

  return (
    <ShoppingContext.Provider value={parsed}>
      {children}
    </ShoppingContext.Provider>
  );
};

const useShopping = () => {
  const context = React.useContext(ShoppingContext);

  if (context === undefined) {
    throw new Error(`useShopping must be used within a ShoppingProvider`);
  }

  return context;
};

export { ShoppingProvider, useShopping };
