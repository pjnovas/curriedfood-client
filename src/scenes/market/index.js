export { ShoppingListScreen } from './shopping-list';
export { IngredientEdit } from './ingredient-edit';
export { IngredientNew } from './ingredient-new';
