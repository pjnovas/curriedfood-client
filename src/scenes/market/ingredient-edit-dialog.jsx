import React, { useState } from 'react';
import { StyleSheet } from 'react-native';
import { Portal, Dialog, Button } from 'react-native-paper';
import get from 'lodash/get';
import Theme from 'theme';
import Chips from 'components/chips';

export const IngredientEdit = ({ onDismiss, onSubmit, allTags, ...item }) => {
  const [tags, setValue] = useState(get(item, 'product.shop_tags', ''));

  return (
    <Portal>
      <Dialog style={styles.layout} visible dismissable onDismiss={onDismiss}>
        <Dialog.Title style={styles.title}>{item.product.name}</Dialog.Title>
        <Dialog.Content style={styles.content}>
          <Chips
            list={allTags}
            value={tags}
            onChange={(value) => {
              setValue(value);
              console.log(value);
            }}
          />
        </Dialog.Content>
        <Dialog.Actions style={styles.actions}>
          <Button onPress={onDismiss}>Cancelar</Button>
          <Button
            dark
            mode="contained"
            style={styles.okButton}
            onPress={() => onSubmit(tags)}
          >
            OK
          </Button>
        </Dialog.Actions>
      </Dialog>
    </Portal>
  );
};

const styles = StyleSheet.create({
  layout: {},
  title: {
    color: Theme.colors.text,
    textAlign: 'center'
  },
  content: {
    minHeight: 100,
    maxHeight: 400
  },
  actions: {},
  okButton: {
    marginHorizontal: 10
  }
});

// export const useFormMarket = () => ({
//   form: useForm()
// });

// export default composeHooks({ useFormMarket })(IngredientEdit);
export default IngredientEdit;
