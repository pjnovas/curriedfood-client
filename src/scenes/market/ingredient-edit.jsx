import React from 'react';
import { StyleSheet } from 'react-native';
import { Formik } from 'formik';
import { FAB } from 'react-native-paper';

import find from 'lodash/find';
import get from 'lodash/get';

import { useRequestSWR } from 'hooks/service';
import { useNavigateBack } from 'hooks/navigation';
import { SafeAreaLayout, SaveAreaInset } from 'components/safe-area-layout';

import { FormInput } from 'components/form-input';
import { Toolbar } from 'components/toolbar';
import Spinner from 'components/spinner';
import Layout from 'components/layout';

// TODO: should check cache and fetch by id (web mode)

export const IngredientEdit = ({
  route: {
    params: { id, title }
  }
}) => {
  const { data } = useRequestSWR(
    {
      url: `/shopping-lists`
    },
    { withPlace: true }
  );

  const item = data && find(get(data, '[0].groceries'), { id });

  const renderForm = ({ handleSubmit }) => (
    <>
      <FormInput id="quantity" style={styles.formControl} label="Cantidad" />
      <FAB style={styles.fab} icon="save" onPress={handleSubmit} />
    </>
  );

  return (
    <SafeAreaLayout style={styles.safeArea} insets={SaveAreaInset.TOP}>
      <Toolbar title={`Editar ${title}`} onBackPress={useNavigateBack()} />
      {!item ? (
        <Spinner bar />
      ) : (
        <Layout style={styles.form}>
          <Formik
            initialValues={item}
            // validationSchema={}
            onSubmit={(values) => console.log(values)}
          >
            {renderForm}
          </Formik>
        </Layout>
      )}
    </SafeAreaLayout>
  );
};

const styles = StyleSheet.create({
  safeArea: {
    flex: 1,
    paddingBottom: 0,
    position: 'relative'
  },
  form: {
    padding: 20
  },
  formContainer: {
    flex: 1,
    paddingVertical: 16,
    paddingHorizontal: 16
  },
  fab: {
    position: 'absolute',
    margin: 16,
    right: 0,
    bottom: 0
  }
});

export default IngredientEdit;
