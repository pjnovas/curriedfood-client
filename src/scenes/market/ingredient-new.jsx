import React from 'react';
import { Text, StyleSheet } from 'react-native';
import { useNavigateBack } from 'hooks/navigation';

import { SafeAreaLayout, SaveAreaInset } from 'components/safe-area-layout';
import { Toolbar } from 'components/toolbar';
import Layout from 'components/layout';

export const IngredientNew = () => (
  <SafeAreaLayout style={styles.safeArea} insets={SaveAreaInset.TOP}>
    <Toolbar title="Nuevo Ingrediente" onBackPress={useNavigateBack()} />
    <Layout>
      <Text>Nuevo Ingrediente</Text>
    </Layout>
  </SafeAreaLayout>
);

const styles = StyleSheet.create({
  safeArea: {
    flex: 1,
    paddingBottom: 0,
    position: 'relative'
  }
});

export default IngredientNew;
