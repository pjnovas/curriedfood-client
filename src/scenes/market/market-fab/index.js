import FabButton from './fab-button';

export default FabButton;
