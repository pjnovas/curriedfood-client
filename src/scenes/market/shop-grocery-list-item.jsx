import React from 'react';
import { ListItem, Text } from '@ui-kitten/components';

const unitText = {
  un: '',
  cu: 'cucharadas de',
  gr: 'gramos de',
  lt: 'litros de',
  tz: 'tazas de',
  ml: 'mililitros'
};

const getText = ({
  product: { name },
  quantity,
  unit: { code },
  alt_quantity,
  alt_unit
}) => {
  let text = `${quantity} ${unitText[code] || ''} ${name}`;

  if (alt_quantity) {
    text += ` (${alt_quantity} ${alt_unit.code})`;
  }

  return text;
};

const ShopGroceryListItem = (props) => (
  <ListItem {...props}>
    <Text category="s1">{getText(props)}</Text>
  </ListItem>
);

export default ShopGroceryListItem;
