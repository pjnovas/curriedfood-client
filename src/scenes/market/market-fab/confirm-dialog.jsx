import React from 'react';
import PropTypes from 'prop-types';
import { StyleSheet } from 'react-native';
import { Paragraph } from 'react-native-paper';
import { composeHooks } from 'utils/language';

import { useMarket } from 'hooks/market';
import Confirm from 'components/confirmation';
import Theme from 'theme';

export const ConfirmDialog = ({ ingredients, ...props }) => (
  <Confirm
    visible
    title="Enviar al carrito de compras?"
    okText="Enviar"
    {...props}
  >
    <Paragraph
      style={styles.contentText}
    >{`Se agregar√°n ${ingredients.length} Ingredientes`}</Paragraph>
  </Confirm>
);

ConfirmDialog.propTypes = {
  ...Confirm.PropTypes,
  ingredients: PropTypes.array
};

ConfirmDialog.defaultProps = {
  ingredients: []
};

const styles = StyleSheet.create({
  contentText: {
    color: Theme.colors.text,
    textAlign: 'center'
  }
});

export const useConfirm = ({ ingredients }) => {
  const [state, { addFrom }] = useMarket();

  return {
    onConfirm: () => addFrom(ingredients),
    loading: state.status !== 'idle'
  };
};

export default composeHooks({ useConfirm })(ConfirmDialog);
