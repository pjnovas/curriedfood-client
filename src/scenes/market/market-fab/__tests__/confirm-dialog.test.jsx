import { useMarket } from 'hooks/market';
import { ConfirmDialog, useConfirm } from '../confirm-dialog';

jest.mock('hooks/market');

const ingredients = [
  {
    id: 1,
    quantity: 3,
    product: {
      name: 'Eggs',
      divisible: false,
      unit: 'un'
    }
  },
  {
    id: 2,
    quantity: 150,
    product: {
      name: 'Olives',
      divisible: true,
      unit: 'gr'
    }
  }
];

takeSnapshots(
  {
    default: {},
    withIngredients: {
      ingredients
    },
    loading: {
      ingredients,
      loading: true
    }
  },
  ConfirmDialog
);

describe('useConfirm', () => {
  const addFrom = jest.fn();
  const mockMarket = (status) => {
    useMarket.mockImplementation(() => [{ status }, { addFrom }]);
  };

  beforeEach(() => {
    addFrom.mockClear();
    useMarket.mockReset();
  });

  test('should return a default state', () => {
    mockMarket('idle');
    const result = useConfirm({ ingredients });

    expect(result).toEqual(
      expect.objectContaining({
        loading: false
      })
    );

    expect(addFrom).not.toBeCalled();
    result.onConfirm();
    expect(addFrom).toBeCalledWith(ingredients);
  });

  test('should return a loading state', () => {
    mockMarket('fetching');
    const result = useConfirm({ ingredients });

    expect(result).toEqual(
      expect.objectContaining({
        loading: true
      })
    );
  });
});
