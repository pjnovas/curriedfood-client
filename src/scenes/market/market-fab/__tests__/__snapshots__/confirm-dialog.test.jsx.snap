// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`default 1`] = `
<Confirmation
  okText="Enviar"
  title="Enviar al carrito de compras?"
  visible={true}
>
  <Paragraph
    style={
      Object {
        "color": "#ffffff",
        "textAlign": "center",
      }
    }
  >
    Se agregarán 0 Ingredientes
  </Paragraph>
</Confirmation>
`;

exports[`loading 1`] = `
<Confirmation
  loading={true}
  okText="Enviar"
  title="Enviar al carrito de compras?"
  visible={true}
>
  <Paragraph
    style={
      Object {
        "color": "#ffffff",
        "textAlign": "center",
      }
    }
  >
    Se agregarán 2 Ingredientes
  </Paragraph>
</Confirmation>
`;

exports[`withIngredients 1`] = `
<Confirmation
  okText="Enviar"
  title="Enviar al carrito de compras?"
  visible={true}
>
  <Paragraph
    style={
      Object {
        "color": "#ffffff",
        "textAlign": "center",
      }
    }
  >
    Se agregarán 2 Ingredientes
  </Paragraph>
</Confirmation>
`;
